src/browser/dom/tree_walker.zig

```diff
@@ -0,0 +1,187 @@
+            .current_node = node,
+            .what_to_show = what_to_show orelse NodeFilter._SHOW_ALL,
+            .filter = filter_func,
+            .depth = 0,
+        };
+    }
+
+    fn verify_what_to_show(self: *const TreeWalker, node: *parser.Node) !bool {
+        const node_type = try parser.nodeType(node);
+        const what_to_show = self.what_to_show;
+        return switch (node_type) {
+            .attribute => what_to_show & NodeFilter._SHOW_ATTRIBUTE != 0,
+            .cdata_section => what_to_show & NodeFilter._SHOW_CDATA_SECTION != 0,
+            .comment => what_to_show & NodeFilter._SHOW_COMMENT != 0,
+            .document => what_to_show & NodeFilter._SHOW_DOCUMENT != 0,
+            .document_fragment => what_to_show & NodeFilter._SHOW_DOCUMENT_FRAGMENT != 0,
+            .document_type => what_to_show & NodeFilter._SHOW_DOCUMENT_TYPE != 0,
+            .element => what_to_show & NodeFilter._SHOW_ELEMENT != 0,
+            .entity => what_to_show & NodeFilter._SHOW_ENTITY != 0,
+            .entity_reference => what_to_show & NodeFilter._SHOW_ENTITY_REFERENCE != 0,
+            .notation => what_to_show & NodeFilter._SHOW_NOTATION != 0,
+            .processing_instruction => what_to_show & NodeFilter._SHOW_PROCESSING_INSTRUCTION != 0,
+            .text => what_to_show & NodeFilter._SHOW_TEXT != 0,
+        };
+    }
+
+    fn verify_filter(self: *const TreeWalker, node: *parser.Node) !bool {
+        if (self.filter) |f| {
+            const filter = try f.call(u32, .{node});
+            return switch (filter) {
+                NodeFilter._FILTER_ACCEPT => true,
+                else => false,
+            };
+        } else return true;
+    }
+
+    pub fn get_root(self: *TreeWalker) *parser.Node {
+        return self.root;
+    }
+
+    pub fn get_currentNode(self: *TreeWalker) *parser.Node {
+        return self.current_node;
+    }
+
+    pub fn get_whatToShow(self: *TreeWalker) u32 {
+        return self.what_to_show;
+    }
+
+    pub fn get_filter(self: *TreeWalker) ?Env.Function {
+        return self.filter;
+    }
+
+    pub fn _firstChild(self: *TreeWalker) !?*parser.Node {
+        const children = try parser.nodeGetChildNodes(self.current_node);
+        const child_count = try parser.nodeListLength(children);
+
+        for (0..child_count) |i| {
+            const index: u32 = @intCast(i);
+            const child = (try parser.nodeListItem(children, index)) orelse return null;
+
+            if (!try self.verify_what_to_show(child)) continue;
+            if (!try self.verify_filter(child)) continue;
+
+            self.depth += 1;
+            self.current_node = child;
+            return child;
+        }
+
+        return null;
+    }
+
+    pub fn _lastChild(self: *TreeWalker) !?*parser.Node {
+        const children = try parser.nodeGetChildNodes(self.current_node);
+        const child_count = try parser.nodeListLength(children);
+
+        for (0..child_count) |i| {
```

Comment:
I was confused why `lastChild` was walking through the list in this order. Now I see the `child_count - 1 - i`.

It's fine as-is, but I think this alternative is the same and might be more explicit:

```zig
var index: u32 = child_count;
while (index > 0) {
   index -= 1
  ...
```